packages/jss-plugin-syntax-rule-value-function/src/index.js

```diff
@@ -13,51 +22,59 @@ export default function functionPlugin() {
     onCreateRule(name: string, decl: JssStyle, options: RuleOptions): Rule | null {
       if (typeof decl !== 'function') return null
       const rule: StyleRuleWithRuleFunction = (createRule(name, {}, options): any)
-      rule[fnStyleNs] = decl
+      rule[fnRuleNs] = decl
       return rule
     },
 
     onProcessStyle(style: JssStyle, rule: Rule): JssStyle {
-      const fn = {}
+      const fnValues = {}
       for (const prop in style) {
         const value = style[prop]
         if (typeof value !== 'function') continue
         delete style[prop]
-        fn[prop] = value
+        fnValues[prop] = value
       }
-      // $FlowFixMe: Flow complains...
-      rule[fnValuesNs] = fn
+      // $FlowFixMe
+      rule[fnValuesNs] = fnValues
       return style
     },
 
-    onUpdate(data: Object, rule: Rule) {
+    onUpdate(data: Object, rule: Rule, sheet: StyleSheet, options: UpdateOptions) {
       // It is a rules container like for e.g. ConditionalRule.
       if (rule.rules instanceof RuleList) {
-        rule.rules.update(data)
+        rule.rules.update(data, options)
         return
       }
 
-      if (rule && rule.type !== 'style') return
-
       const styleRule: StyleRule = (rule: any)
 
-      // If we have a fn values map, it is a rule with function values.
       // $FlowFixMe
-      if (styleRule[fnValuesNs]) {
-        for (const prop in styleRule[fnValuesNs]) {
-          styleRule.prop(prop, styleRule[fnValuesNs][prop](data))
+      const fnRule = styleRule[fnRuleNs]
+
+      // If we have a style function, the entire rule is dynamic and style object
+      // will be returned from that function.
+      if (fnRule) {
+        styleRule.style = fnRule(data)
+
+        // We need to run the plugins in case style relies on syntax plugins.
+        const process = options ? options.process : true
+        if (process) {
+          styleRule.options.jss.plugins.onProcessStyle(styleRule.style, styleRule, sheet)
```

Comment:
don't really like this here, that api was supposed to be private, but  I haven't found a way around so far.

Reply:
Can't this be moved into the core and then diff it against the previous style object and only update the properties which are needed?

Reply:
Also I thought we would go with the route of `fastValue((props) => props.color)` to make fast function values?

Reply:
strict comparison of previous and current style reference before the hook and after might actually be a good idea! Gonna try this:

`fastValue` we can still add as a separate feature later on, I think we can skip this for now, just because we still have observables for "fast by default" and its not clear if anyone actually using function values as "fast by default" besides of my demos.

Reply:
also  `sheet.update(data, {process: true})` is an equevalent, so we might not need `fastValue()`  at all

Reply:
> We could also make a shallow comparison right?

Yeah, can be done if needed, I think we only need to check style references currently.

> Also what do you think of only supporting fn values as fast and not fn rules?

The problem is that values with jss-expand can be complex too, for e.g. `margin: {top, left}`, so you will have a painful inconsistency if you make them fast by default.

Reply:
No. not make them fast by default but instead only allowing 'fast' functions as function values.

Reply:
same thing, no? `margin: () => (top, left)`